pytest_httpserver/httpserver.py

```diff
@@ -663,7 +664,19 @@ def url_for(self, suffix: str):
         else:
             protocol = "https"
 
-        return "{}://{}:{}{}".format(protocol, self.host, self.port, suffix)
+        try:
+            if (
+                isinstance(ipaddress.ip_address(self.host), ipaddress.IPv6Address)
```

Comment:
This seems to be the proper way to check, but I'm thinking about whether it would be simpler to have:

```python
if ":" in self.host and not self.host.startswith("[") and not self.host.endswith("]"):
    host = f"[{self.host]"
else:
    host = self.host
```

So the code would be a bit more readable by avoiding try-except block.
What is a question if it is 100% to be an IPv6 address if it contains a ":" and it is 100% a non-ipv6 if it does not.
I think it is true but this needs to be verified.


I am also questioning the API breaking, eg. if someone now uses IPv6 and calls `url_for` and then manipulates the result by adding the square brackets, then this will (or could...) break their code.

Let me think about this.
Maybe an additional `url_for_ipv6` or something else would be needed to avoid this but I'm unsure.

Reply:
I have a better idea :)

Subclass from `HTTPServer`, create `HTTPServerIPv6` subclass and override `url_for` method so it will always add the square brackets (it can assume that `host` is an IPv6 address, we can add additional checks for this in the `__init__`). `httpserver_ipv6` would use this subclass. So at the end we can keep the original behavior for the `HTTPServer`.

Otherwise, if you think that is is complicated, we can also break the API, so adding this conditional formatting to the `url_for`, and document it, maybe look at github if anyone uses it in a way we could break it.

What do you think?

Reply:
> What is a question if it is 100% to be an IPv6 address if it contains a ":" and it is 100% a non-ipv6 if it does not.

Honestly I don't know, this is why I relied on the ipaddress python module to detect IPV6 :)

> I am also questioning the API breaking, eg. if someone now uses IPv6 and calls url_for and then manipulates the result by adding the square brackets, then this will (or could...) break their code.

This is true. However:
- this would not concern a majority of users
- this is quite easy to fix
- I guess `pytest-httpserver` in only used in unit test contexts, not in production context. This makes little breaking changes more acceptable to me.

IMHO the solutions you suggest (`url_for_ipv6` or `HTTPServerIPV6`) adds too much complexity in regard of the feature needed (i.e. always generate valid urls). I agree that the `try/except` blocks decreases the readability, so I would suggest:

- call a dedicated function to generate urls from `url_for`:
  - either use a dedicated tool like [furl](https://github.com/gruns/furl) and rely on it to always generate a valid URL
  - or extract the current code I wrote in a custom `make_url` function
- go for the API break, and document it

Does this sound acceptable to you?

Reply:
Thanks for sharing your thoughts!

I accept that we should update it in `url_for`. We will document it in the changelist and probably bump the 2nd version (eg. release 1.1.0). I think you are right that it is quite easy to fix and we will break the tests only, not the production code.

I'm also ok with using ipaddress module, I would not bring in a 3rd party library unless absolutely needed. I looked the source of ipaddress.py in cpython in the hope we could extract some logic but it does not seem to be easy so I agree that we should use the correct tool to identify the version of the IP address.

I think readability can be made better like this:
```python
        try:
            ip_address = ipaddress.ip_address(self.host)
        except ValueError:
            host = self.host
        else:
            assert ip_address.version in (4, 6)
            if ip_address.version == 4:
                host = self.host
            elif ip_address.version == 6:
                if not self.host.startswith("[") and not self.host.endswith("]"):
                    host = f"[{self.host}]"
                else:
                    host = self.host
```

This uses the else branch of the try..except, which is not widely known but if it is a problem, a `host = None` can be added before try and then check for `if host is None:` later.

I think what I don't like in your implementation is that try..except block is too long and for me it was hard to figure out which function may raise ValueError, and why we are handling it (then I realized..).

I don't want to specify the exact implementation to you so it's your call. If you think that the current version is ok, then I can also accept this PR. If you find it you can improve it on readability then it is better but not a must.

Let me know.. :)